README.md
# PictoWeb
PictoChat clone using socket.io


CHANGELOG.md
# Changelog

## 0.1.2 (2021-3-1)
### Added
- Experimental mobile support
- GitHub corner link

## 0.1.1 (2020-12-26)
### Fixed
- Fixed page scroll glitching
- Fixed tooltip remaining after selection
- Better screen responsiveness
- Improved alignment of certain elements

### Added
- Added Changelog

## 0.1
- Initial Release

server.js
const path = require('path');
const http = require('http');
const express = require('express');
const socketio = require('socket.io');

const formatMssg = require('./public/utils/messages');
const formatColor = require('./public/utils/colors');
const {userJoin, getCurrentUser} = require('./public/utils/users');

const app = express();
const server = http.createServer(app);
const io = socketio(server);

//static folder (html)
app.use(express.static(path.join(__dirname, 'public')));

//run on client connection
io.on('connection', (socket)=>{

    socket.on('joinChat', ({username, color})=> {
        const user = userJoin(socket.id, username, color);

        socket.join();

        //user welcome message
        socket.emit('mssg', 'Welcome to PictoWeb!');
        socket.emit('canvasProp', getCurrentUser(socket.id), formatColor(user.color));

        //broadcast on user connection
        socket.broadcast.emit('mssg', `Now entering: ${user.username}`);
    });

    //listen for emit canvasMssg
    socket.on('canvasMssg', (mssg)=>{
        const user = getCurrentUser(socket.id);
        io.emit('canvasMssg', formatMssg(user.username, user.color, mssg));
    });

    //run on client disconnect
    socket.on('disconnect', ()=>{
        const user = getCurrentUser(socket.id);
        io.emit('mssg', `Now leaving: ${user.username}`);
    });
});

const PORT = process.env.PORT || 3000;

server.listen(PORT, ()=> console.log(`server running on port ${PORT}`));

public/js/main.js
const sendBttn = document.getElementById('send');
const dataHolder = document.getElementById('canvasDataHolder');
const mssgDiv = document.getElementById('containerMssg');
// const mssgName = document.querySelector('.mssgUsername');

const canvasEl = document.getElementById('drawLayer');
const usernameEl = document.getElementById('name');
const foot = document.getElementById('footer');
var footHidden = false;

const dropHeight = document.body.offsetHeight - document.getElementById('containerHeader').offsetHeight - document.getElementById('mobileHideUserElements').offsetHeight + 13;
const dropStr = dropHeight + "px";
const origStr = mssgDiv.offsetHeight + "px";
mssgDiv.style.height = origStr;
foot.style.height = foot.offsetHeight + "px";
// const {userJoin, getCurrentUser} = require('../utils/users');
// const formatColor = require('../utils/colors');


const socket = io();


// Get username and color from url
const {username, color} = Qs.parse(location.search, {
    ignoreQueryPrefix: true
})

// User joins the chat
socket.emit('joinChat', {username, color});

socket.on('canvasProp', (currentUser, userColor)=>{
    const user = currentUser;
    // console.log(canvasEl, usernameEl);
    canvasEl.setAttribute('style', `border-color: ${userColor.primary}`);
    usernameEl.setAttribute('style', `background-color: ${userColor.secondary};
                                        border-color: ${userColor.primary};
                                        color: ${userColor.primary}`);
    usernameEl.innerHTML = user.username;
    // usernameEl.setAttribute('style', ``);
    // usernameEl.setAttribute('style', ``);
});

// Mesages from server
socket.on('mssg', (mssg)=>{
    outputMessage(mssg);
    mssgDiv.scrollTop = mssgDiv.scrollHeight;
});
// Messages from user
socket.on('canvasMssg', (mssg)=>{
    outputCanvasMssg(mssg);
    // formatNameStyle(mssg);
    mssgDiv.scrollTop = mssgDiv.scrollHeight;
});

// Send message
$(document).on('canvasSrcAdded', (e)=>{
    // get canvas data once toDataURL() completes
    const mssgData = dataHolder.getAttribute('src');
    
    // emit message (canvas data) to server
    socket.emit('canvasMssg', mssgData); 
});

// Output message to DOM (#containerMssg)
function outputCanvasMssg(mssg){
    const div = document.createElement('div')
    div.classList.add('mssg');
    div.setAttribute('style', `border-color: ${mssg.color.primary}`);
    
    div.innerHTML = `<p class='h6 disableSelect mssgUsername' 
                        style='border-color: ${mssg.color.primary};
                        background-color: ${mssg.color.secondary};
                        color: ${mssg.color.primary}'>
                        ${mssg.username}
                    </p>
                    <img src="${mssg.mssg}">`;
    // $(div).html(`<p class='h6 disableSelect mssgUsername'>${mssg.username}</p>
    //             <img src="${mssg.mssg}">`);
    
    // console.log(mssgName);

    mssgDiv.appendChild(div);

    // mssgName.setAttribute('style', `border-color: ${mssg.color.primary}`);
    // mssgName.setAttribute('style', `background-color: ${mssg.color.secondary}`);
}
// function formatNameStyle(mssg){
//     mssgName.setAttribute('style', `border-color: ${mssg.color.primary}`);
//     mssgName.setAttribute('style', `background-color: ${mssg.color.secondary}`);
// }
function outputMessage(mssg){
    const div = document.createElement('div')
    div.classList.add('sysMssg');
    div.classList.add('disableSelect');
    div.innerHTML = `<p class='sysText'>${mssg}</p>`;
    mssgDiv.appendChild(div);
}

function hidePalette(){
    if (!footHidden) {
        foot.style.height = "0";
        mssgDiv.style.height = dropStr;
        footHidden = true;
    } else {
        foot.style.height = "200px";
        mssgDiv.style.height = origStr;
        footHidden = false;
    }
}

public/js/textLayer.js
window.addEventListener("load", ()=>{
    const canvas = document.querySelector("#textLayer");
    const ctx = canvas.getContext("2d");

    if(canvas.getContext){
        // //window size
        // canvas.width = 450;
        // canvas.height = 160;

        // var w = canvas.getBoundingClientRect().width - 450;
        // var h = canvas.getBoundingClientRect().height - 160;
        // canvas.width = canvas.getBoundingClientRect().width - w;
        // canvas.height = canvas.getBoundingClientRect().height - h;
        resize();

        
        //initial variables
        var width = document.querySelector('#name').borderWidth;
        var color = window.getComputedStyle(document.getElementById('name'), "").getPropertyValue("background-color");
        // ctx.imageSmoothingEnabled = !1;
        // ctx.filter = 'url(#remove-alpha)';
        ctx.translate(0.5, 0.5);

        const userColors = [
                    ["#B8B8F0", "#121592"],//ppl
                    ["#FEA1A1", "#AF2121"],//red
                    ["#8CF1BA", "#0d793d"],//grn
                    ["#feb588", "#e34400"],//org
                    ["#76dff2", "#086e9a"]//blu
                    ];

        function resize() {
            var w = canvas.getBoundingClientRect().width - 450;
            var h = canvas.getBoundingClientRect().height - 160;
            canvas.width = canvas.getBoundingClientRect().width - w;
            canvas.height = canvas.getBoundingClientRect().height - h;
        }

        

        for(let i = 32; i < 160; i += 32){
            ctx.beginPath();
            ctx.strokeStyle = color;
            ctx.moveTo(0, i);
            ctx.lineTo(canvas.width, i);
            ctx.stroke();
        }
    }
});

public/js/userColor.js
const userColors = [
    ["#B8B8F0", "#121592"],//ppl
    ["#FEA1A1", "#AF2121"],//red
    ["#8CF1BA", "#0d793d"],//grn
    ["#feb588", "#e34400"],//org
    ["#76dff2", "#086e9a"]//blu
    ];

    for (let i = 0; i < 5; i++) {
        var c1 = userColors[i][0];
        var c2 = userColors[i][1]
        $(".c"+(i+1).toString()).css("backgroundColor", c1)
                                .css("borderColor", c2)
                                .css("borderWidth", 1);
        
        // $(".c"+(i+1).toString()+":active")
        //     .css("backgroundColor", "red")
        //     .css("borderColor", "#red");
        // $(".c"+(i+1).toString()+":focus")
        //     .css("backgroundColor", "red")
        //     .css("borderColor", "#red");
        // $(".c"+(i+1).toString()+".focus")
        //     .css("backgroundColor", "red")
        //     .css("borderColor", "#red");
    }

    var group = $('.rbColor');

    group.change(function(){ //preserves visisble checked state
        var selected = "."+$('input[name=color]:checked').prop('id');
        // console.log(selected);
        group.css("borderWidth", 1);
        $(selected).css("borderWidth", 2);
    });

    $(".c1").button('toggle'); //default


public/js/canvas.js

window.addEventListener("load", ()=>{
    const canvas = document.querySelector("#drawLayer");
    const ctx = canvas.getContext("2d");


    if(canvas.getContext){
        // //window size
        // canvas.width = 450;
        // canvas.height = 160;

        // var w = canvas.getBoundingClientRect().width - 450;
        // var h = canvas.getBoundingClientRect().height - 160;
        // canvas.width = canvas.getBoundingClientRect().width - w;
        // canvas.height = canvas.getBoundingClientRect().height - h;
        resize();

        //create line dividers
        // ctx.strokeStyle = document.getElementById('name').getAttribute('background-color');
        // for(let i = 0; i < 160; i=i+32){
        //     ctx.moveTo(0, i);
        //     ctx.lineTo(canvas.width, i);
        //     ctx.stroke();
        // }

        let paint = false;

        //initial variables
        var width = 5;
        var color = "#000000";
        ctx.imageSmoothingEnabled = !1;
        ctx.filter = 'url(#remove-alpha)';

        // function startPos(e){
        //     paint = true;
        //     draw(e);
        // }
        // function endPos(e){
        //     paint = false;
        //     ctx.beginPath();
        // }



        // last known position
        var pos = { x: 0, y: 0 };

        function resize() {
            var w = canvas.getBoundingClientRect().width - 450;
            var h = canvas.getBoundingClientRect().height - 160;
            canvas.width = canvas.getBoundingClientRect().width - w;
            canvas.height = canvas.getBoundingClientRect().height - h;
            
            // var name = document.querySelector('#name'); 
            // if(name.offsetHeight <= 35){
            //     // name.style.paddingBottom = '1px';
            // }
        }

        // new position from mouse event
        function setPosition(e) {
            const rect = canvas.getBoundingClientRect();
            pos.x = e.clientX - rect.left;
            pos.y = e.clientY - rect.top;
        }

        // new position from touch event
        function setPositionTouch(e){
            try{
                const rect = canvas.getBoundingClientRect();
                pos.x = e.touches[0].clientX - rect.left;
                pos.y = e.touches[0].clientY - rect.top;
            } catch(err){}
        }
        
        function drawNew(e) {
            // mouse left button must be pressed
            if (e.buttons !== 1) return;
          
            ctx.beginPath(); // begin
          
            ctx.lineWidth = width;
            ctx.strokeWidth = width;
            ctx.lineCap = 'round';
            ctx.lineJoin = 'round';
            ctx.strokeStyle = color;
            
            ctx.moveTo(pos.x, pos.y); // from
            setPosition(e);
            ctx.lineTo(pos.x, pos.y); // to
          
            ctx.stroke();
        }

        function drawNewTouch(e) {
            ctx.beginPath(); // begin
          
            ctx.lineWidth = width;
            ctx.strokeWidth = width;
            ctx.lineCap = 'round';
            ctx.lineJoin = 'round';
            ctx.strokeStyle = color;

            ctx.moveTo(pos.x, pos.y); // from
            setPositionTouch(e);
            ctx.lineTo(pos.x, pos.y); // to
          
            ctx.stroke();
        }

        // function draw(e){
        //     if(!paint) return;

        //     const rect = canvas.getBoundingClientRect();
        //     const x = MouseEvent.clientX - rect.left;
        //     const y = MouseEvent.clientY - rect.top;

        //     ctx.lineWidth = width;
        //     ctx.strokeStyle = color;
        //     ctx.lineCap = "round";

        //     ctx.lineTo(x, y);
        //     ctx.stroke();
        //     ctx.beginPath();
        //     ctx.moveTo(x, y);
            
        // }

        function lineWidth(c){
            switch(c){
                case 'lg':
                    width = 12;
                    break;
                case 'md':
                    width = 5;
                    break;
                case 'sm':
                    width = 2;
                    break;
                default:
                    break;
            }
        }
        function clear(){
            ctx.clearRect(0,0,canvas.width, canvas.height);
            //document.getElementById("canvas").style.display = "none";
        }
        
        function erase(toggle){
            if(toggle){
                ctx.globalCompositeOperation="destination-out";
                color = "#ffffff";
            } else {
                ctx.globalCompositeOperation="source-over";
                color = "#000000";
            }
        }


        function send(){
            var canvasData = canvas.toDataURL();
            document.getElementById('canvasDataHolder').setAttribute("src", canvasData);
            $(document).trigger('canvasSrcAdded');
        }



        //event listeners
        // window.addEventListener('resize', resize);
        // canvas.addEventListener("mouseup", endPos);


        // Canvas desktop mouse event listeners
        canvas.addEventListener("mousedown", setPosition, false);
        canvas.addEventListener("mousemove", drawNew, false);
        canvas.addEventListener("mouseenter", setPosition, false);
        canvas.addEventListener("click", function(){
            ctx.beginPath();
            ctx.fillStyle = color;
            ctx.arc(pos.x, pos.y, width/2, 0, 2 * Math.PI);
            ctx.fill();
        }, false);

        // Canvas mobile touch event listeners
        canvas.addEventListener("touchstart", setPositionTouch, false);
        canvas.addEventListener("touchmove", drawNewTouch, false);
        canvas.addEventListener("touchenter", setPositionTouch, false);
        canvas.addEventListener("touchend", setPositionTouch, false);



        // $('canvas').mousemove(function(e){
        //     if(e.buttons == 1) console.log("y: " + pos.y);
        // });
        // $('canvas').click(()=> console.log("y: " + pos.y));

        document.getElementById("lg").addEventListener('click', ()=> lineWidth('lg'));
        document.getElementById("md").addEventListener('click', ()=> lineWidth('md'));
        document.getElementById("sm").addEventListener('click', ()=> lineWidth('sm'));
        document.getElementById("clear").addEventListener('click', ()=>{
            erase(false)
            clear()
        });
        document.getElementById("erase").addEventListener('click', ()=> erase(true));
        document.getElementById("draw").addEventListener('click', ()=> erase(false));
        document.getElementById("send").addEventListener('click', ()=>{
            send()
            clear()
        });
        // document.getElementById("clone").addEventListener('click', ()=>{
        // });
    }
});